clang/runtime/dpct-rt/include/dpl_extras/algorithm.h

```diff
@@ -1455,8 +1455,604 @@ inline void segmented_sort_pairs_by_two_pair_sorts(
+        [=](sycl::nd_item<1> __self_item) {
+          const ::std::size_t __self_lidx = __self_item.get_local_id(0);
+          const ::std::size_t __wgroup_idx = __self_item.get_group(0);
+          const ::std::size_t __seg_start =
+              WorkGroupSize * iters_per_work_item * __wgroup_idx;
+
+          __clear_wglocal_histograms(hacc_private, __wgroup_idx * num_bins,
+                                     num_bins, __self_item);
+          __accum_local_atomics<WorkGroupSize, ::std::uint32_t,
+                                sycl::access::address_space::global_space>(
+              input, __seg_start, N, __self_lidx, hacc_private,
+              __wgroup_idx * num_bins, func, __self_item, iters_per_work_item);
+
+          __reduce_out_histograms<BinType>(hacc_private,
+                                           __wgroup_idx * num_bins, bins,
+                                           num_bins, __self_item);
+        });
+  });
+  e.wait();
+}
+
+template <typename Policy, typename Iter1, typename Iter2, typename Size,
+          typename IdxHashFunc, typename... Range>
+inline Iter2
+__histogram_general_select_best(Policy &&policy, Iter1 first, Iter1 last,
+                                Iter2 histogram_first, const Size &num_bins,
+                                IdxHashFunc func, Range &&...opt_range) {
+  using __histo_value_type = typename ::std::iterator_traits<Iter2>::value_type;
+  auto __local_mem_size =
+      policy.queue()
+          .get_device()
+          .template get_info<sycl::info::device::local_mem_size>();
+  constexpr ::std::uint8_t __max_registers = 16;
+
+  // Temporary use of stable non-public API from oneDPL,  this function will be
+  // replaced with oneDPL call in an upcoming release.
+  auto keep_input = oneapi::dpl::__ranges::__get_sycl_range<
+      oneapi::dpl::__par_backend_hetero::access_mode::read, Iter1>();
+  auto input_buf = keep_input(first, last);
+  // Temporary use of stable non-public API from oneDPL,  this function will be
+  // replaced with oneDPL call in an upcoming release.
+  auto keep_bins = oneapi::dpl::__ranges::__get_sycl_range<
+      oneapi::dpl::__par_backend_hetero::access_mode::write, Iter2>();
+  auto bins_buf = keep_bins(histogram_first, histogram_first + num_bins);
+
+  oneapi::dpl::fill(std::forward<Policy>(policy), bins_buf.all_view().begin(),
+                    bins_buf.all_view().end(), __histo_value_type(0));
+  auto N = last - first;
+  if (num_bins < __max_registers) {
+    // If bins fit into registers, use register private accumulation
+    __histogram_general_registers_local_reduction<1024, 32, 16,
+                                                  __histo_value_type>(
+        ::std::forward<Policy>(policy), input_buf.all_view(),
+        bins_buf.all_view(), num_bins, func,
+        std::forward<Range...>(opt_range)...);
+  } else if (num_bins * sizeof(__histo_value_type) < __local_mem_size) {
+    // If bins fit into SLM, use local atomics
```

Comment:
Have you experimented with subgroup size in the implementation that uses SLM?

If not, this can probably be marked as a future optimization, but we may want to consider experimenting with subgroup size of 16 on Intel devices when large amounts of SLM are being used per work-group as this may achieve higher occupancy.

Reply:
I have looked at subgroup size some and have found that intel hardware does do better with 16, but did not see huge differences at the time.  It would be good to revisit it now.  However, I also want this PR to compile without warning on all hardware, so I did not want to specify 16 explicitly.